File: day5.dart
import 'dart:io';

void main() {
    var lines = File('day5.input').readAsLinesSync();
    part1(lines);
    part2(lines);
}

class Range {
    int min=0, max=0;
    Range(this.min, this.max);
    Range.fromString(String s) {
        var list = s.split('-').map(int.parse);
        min = list.first;
        max = list.last;
    }
    bool contains(int i) {
        return min <= i && i <= max;
    }
    bool overlaps(Range other) {
        return other.contains(min) || contains(other.min);
    }
    void merge(Range other) {
        min = (min < other.min) ? min : other.min;
        max = (max > other.max) ? max : other.max;
    }
}

void part1(List lines) {
    var i = lines.indexOf('');
    var ranges = lines.sublist(0,i).map((s) => Range.fromString(s));
    var ids = lines.sublist(i+1).map((id) => int.parse(id));
    print(ids.map((id) => ranges.map((r) => r.contains(id)).reduce((a,b) => a || b)).where((x) => x == true).length);
}

void part2(List lines) {
    var i = lines.indexOf('');
    var ranges = lines.sublist(0,i).map((s) => Range.fromString(s)).toList();
    while (true) {
        List<Range> next = [];
        for (int i=0; i < ranges.length; i++) {
            bool merged = false;
            for (int j=i+1; j < ranges.length; j++) {
                if (ranges[j].overlaps(ranges[i])) {
                    ranges[j].merge(ranges[i]);
                    merged = true;
                }
            }
            if (!merged) {
                next.add(ranges[i]);
            }
        }
        if (ranges.length == next.length) break;
        ranges = next;
    }
    print(ranges.map((r) => r.max-r.min+1).reduce((a,b) => a+b));
}

File: day5.jl
# day5.jl - Advent of Code 2025

function parse_ranges(lines)
    empty = findfirst(==(""), lines)
    lines = lines[1:empty-1]
    [range(a, b) for (a,b) in lines .|> l->split(l, "-") .|> s->parse(Int, s)]
end

function parse_available(lines)
    empty = findfirst(==(""), lines)
    lines = lines[empty+1:end]
    lines .|> s->parse(Int, s)
end

function part1(ranges, available)
    function is_fresh(id)
        any(ranges .|> range->id in range)
    end
    sum(available .|> is_fresh)
end

function part2(ranges, available)
    function overlapping(r1, r2)
        !(r1.stop < r2.start || r1.start > r2.stop)
    end
    function merge(r1, r2)
        min(r1.start, r2.start):max(r1.stop, r2.stop)
    end
    function combine(ranges)
        result = []
        for i in 1:length(ranges)
            merged = false
            for j in i+1:length(ranges)
                if overlapping(ranges[i], ranges[j])
                    ranges[j] = merge(ranges[i], ranges[j])
                    merged = true
                end
            end
            if !merged
                push!(result, ranges[i])
            end
        end
        result
    end
    sum(map(length, combine(ranges)))
end

lines = readlines("day5.input")
ranges = parse_ranges(lines)
available = parse_available(lines)
part1(ranges, available) |> println
part2(ranges, available) |> println
